compiler/src/Parse/Shader.hs

```diff
@@ -29,14 +32,34 @@ import qualified Parse.Primitives.Shader as Shader
 
 shader :: R.Position -> Parser Src.Expr
 shader start@(R.Position row col) =
-  do  block <- Shader.block
-      shdr <- parseSource row col (Text.unpack block)
+  do  startOffset <- getOffset
+      shdr <- parseBlock
+      endOffset <- getOffset
+      block <- getText startOffset endOffset
       end@(R.Position row2 col2) <- getPosition
       let uid = List.intercalate ":" (map show [row, col, row2, col2])
       let src = Text.replace "\n" "\\n" (Text.replace "\r\n" "\\n" block)
       return (A.at start end (Src.Shader (Text.pack uid) src shdr))
 
 
+getText :: Int -> Int -> Parser Text.Text
+getText start end =
+  Parser $ \state@(State fp _ _ _ _ _ _) _ _ eok _ ->
+    let
+      !size = end - start
+      !shader = Text.decodeUtf8 (B.PS fp start size)
+    in
+      eok shader state noError
+
+
+parseBlock :: Parser Shader.Shader
+parseBlock =
+  -- TODO: Parse the `{|glsl`
+  -- TODO: Parse GLSL
+  -- TODO: Parse the `|}`
```

Comment:
This approach might be a bit hard because the GLSL parser needs to check `|}` everywhere. Another approach is to use `Shader.block` to find the offset where GLSL ends, and parse only GLSL using the range found already.

Reply:
I see. In the case of parsing the range we would need to restore the same `row` and `col` in the state.